### PR TITLE
Add detection of [Partial]Ord methods in the `ambiguous_wide_pointer_comparisons` lint

### PR DESCRIPTION
**Pull request by Urgau**
Partially addresses https://github.com/rust-lang/rust/issues/121264 by adding diagnostics items for PartialOrd and Ord methods, detecting such diagnostics items as "binary operation" and suggesting the correct replacement.

I also took the opportunity to change the suggestion to use new methods `.cast()` on `*mut T` an d `*const T`.

### CODE REVIEW

**Comment by Urgau** [comment]
r? compiler

File: compiler/rustc_lint/src/types.rs [hunk 1]
```diff
@@ -873,13 +873,19 @@ impl<'tcx> LateLintPass<'tcx> for TypeLimits {
...
         }
 
-        fn partialeq_binop(diag_item: Symbol) -> Option<hir::BinOpKind> {
-            if diag_item == sym::cmp_partialeq_eq {
-                Some(hir::BinOpKind::Eq)
-            } else if diag_item == sym::cmp_partialeq_ne {
-                Some(hir::BinOpKind::Ne)
-            } else {
-                None
+        fn diag_item_binop(diag_item: Symbol) -> Option<Option<hir::BinOpKind>> {
```
Comment (Nadrieril) [inline compiler/rustc_lint/src/types.rs:876]: `Option<Option<` is not super clear, could you define an `enum ComparisonOp { BinOp(BinOpKind), Other }` or something like that?

File: compiler/rustc_lint/src/types.rs [hunk 2]
```diff
@@ -742,12 +742,12 @@ fn lint_wide_pointer<'tcx>(
...
                     deref_left,
                     deref_right,
-                    // those two Options are required for correctness as having
-                    // an empty span and an empty suggestion is not permitted
-                    left_before: (l_ty_refs != 0).then_some(left),
-                    right_before: (r_ty_refs != 0).then(|| r_span.shrink_to_lo()),
-                    left: l_span.shrink_to_hi(),
-                    right,
+                    paren_left: (l_ty_refs != 0).then_some(")").unwrap_or_default(),
+                    paren_right: (r_ty_refs != 0).then_some(")").unwrap_or_default(),
```
Comment (Nadrieril) [inline compiler/rustc_lint/src/types.rs:746]: Cute, but even longer than the obvious "if" :grin: 
```suggestion
                    paren_left: if l_ty_refs != 0 { ")" } else { "" },
                    paren_right: if r_ty_refs != 0 { ")" } else { "" },
```

**Review by Nadrieril — CHANGES_REQUESTED** [review]

File: compiler/rustc_lint/src/types.rs (hunk 1 above)
Comment (Urgau) [reply compiler/rustc_lint/src/types.rs]: Yeah, that's way better. Thanks.

**Comment by Urgau** [comment]
I've addressed all the review comments and CI passes; this is ready.

@rustbot ready

**Review by Nadrieril — APPROVED** [review]

**Comment by Nadrieril** [comment]
Thanks!

@bors r+

**Comment by Nadrieril** [comment]
One secondary regression on incr alongside some improvements even on the same bench, nothing to worry about

**Pull request merged**